repair/repair.cc

```diff
@@ -1966,23 +1988,57 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
             }
             auto& strat = erm->get_replication_strategy();
             dht::token_range_vector ranges = strat.get_ranges(myid, *tmptr).get0();
-            auto& topology = erm->get_token_metadata().get_topology();
             std::unordered_map<dht::token_range, repair_neighbors> range_sources;
             auto nr_tables = get_nr_tables(db, keyspace_name);
-            rlogger.info("{}: started with keyspace={}, source_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc, ranges.size() * nr_tables, ignore_nodes);
+            sstring ks_source_dc;
+            // Allow repairing in the source_dc only if there are enough replicas remaining
+            if (!source_dc_nodes.empty()) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    const auto& nt_strat = dynamic_cast<const locator::network_topology_strategy&>(strat);
+                    ssize_t rf = nt_strat.get_replication_factor(source_dc);
+                    // Can we ensure a quorum in source_dc?
+                    if (full_racks_in_source_dc.size() > rf / 2) {
+                        ks_source_dc = source_dc;
+                    } else if (reason == streaming::stream_reason::rebuild) {
+                        if (rf == 0) {
+                            throw std::runtime_error(fmt::format("Cannot rebuild with source_dc={}: Replication factor in this data center is 0."
+                                    " Use a different source_dc, or no source_dc option to consider the whole cluster for repair", source_dc));
+                        }
+                        rlogger.warn("Cannot ensure quorum for rebuild with source_dc={}: full racks={} rf={}: rebuilding using source_dc anyway",
```

Comment:
What are full racks?

Reply:
racks with no dead / replaced nodes

Reply:
algorithm changed in v4, does not use full racks anymore